File: app/src/main/java/me/cafecode/android/newspaper/data/NewsRepository.java
package me.cafecode.android.newspaper.data;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import me.cafecode.android.newspaper.data.local.NewsLocalDataSource;
import me.cafecode.android.newspaper.data.models.News;
import me.cafecode.android.newspaper.data.remote.NewsRemoteDataSource;

/**
 * Created by Natthawut Hemathulin on 5/31/2017 AD.
 * Email: natthawut1991@gmail.com
 */

public class NewsRepository implements NewsRepositoryDataSource {

    private NewsRemoteDataSource mRemoteData;
    private NewsLocalDataSource mLocalData;

    @Inject
    public NewsRepository(NewsRemoteDataSource remoteData, NewsLocalDataSource localData) {
        mRemoteData = remoteData;
        mLocalData = localData;
    }


    @Override
    public Observable<List<News>> loadNewses() {
        return mRemoteData.loadNewses();
    }
}

File: app/src/main/java/me/cafecode/android/newspaper/data/NewsRepository.kt
package me.cafecode.android.newspaper.data

import android.util.Log
import io.reactivex.Observable
import me.cafecode.android.newspaper.data.local.NewsLocalDataSource
import me.cafecode.android.newspaper.data.models.News
import me.cafecode.android.newspaper.data.remote.NewsRemoteDataSource
import javax.inject.Inject
import javax.inject.Singleton

@Singleton open class NewsRepository @Inject constructor(remoteData: NewsRemoteDataSource, localData:
NewsLocalDataSource)
    : NewsRepositoryDataSource {

    var mRemoteData = remoteData

    var mLocalData = localData

    val TAG = "NewsRepository"

    override fun loadNewses(): Observable<List<News>> {
        return mRemoteData.loadNewses()
                .flatMap { newses ->
                    Observable.fromArray(newses)
                            .doOnNext { nextNewses ->
                                mLocalData.saveNewses(nextNewses)
                            }
                }
//                .doOnError({ error -> Log.e(TAG, error.message) })
    }

}
